### PR TITLE
[#1451] Cover photo on commodities — A + B + C

### LINKED ISSUE #1451: Cover photo on commodity cards (list + preview)
## Motivation

Today the commodity cards in the React list (`frontend-react/src/pages/commodities/CommoditiesListPage.tsx`) render a flat emoji-by-type icon (🏠 / 💻 / 🔧 / 🪑 / 👕 / 📦 — see `features/commodities/constants.ts:COMMODITY_TYPE_ICONS`). For an inventory product whose value prop is "I never lose track of what I own", a real **photo of the actual item** is the single biggest visual improvement to the list page. The Sheet-preview overlay (#1410) and detail page (#1410) both have the same gap.

Fall back to the type emoji only when the commodity has no photo attached.

## Open question — how is the cover photo selected?

This is research, not a decided design. Options to evaluate as part of this issue:

| Option | Pros | Cons |
|---|---|---|
| **(A)** First image in the legacy `images` (or unified `files` filtered by `category=photos` after #1398) sorted by `created_at` ascending. | Trivial — no schema change, no UI change. Predictable. | User has no control. The first photo they uploaded may be a receipt / box, not the item itself. |
| **(B)** Explicit `cover_file_id` FK on the commodity row, set via a star-icon toggle on each photo in the detail page's Files tab. | User-controllable; clearly expresses intent. | Extra column; must keep referential integrity (clear on file delete). UI for choosing cover. |
| **(C)** Upload-time prompt: "Use as cover photo?" checkbox, defaults to true on the first photo. | Frictionless first-time experience. | Doesn't solve "I want to change it later" — needs (B) anyway. |
| **(D)** ML pick (largest / sharpest / most-centered subject). | Magical. | Way out of scope for self-hosted; opens "wrong pick" failure mode. Reject. |

**Recommended starting position:** ship **(A) as the default + (B) as the override**, with (C) as polish. (B) is a one-column migration plus a small UI affordance and survives any future change in default-pick logic.

Decide as part of this issue:
1. Run the research/spike (sketch (A) and (B) wireframes against the mock) and update this issue with a chosen path.
2. Then break the implementation into sub-issues per (A) and (B), if both ship.

### Backend
- (A) — none. Uses existing `GetImages` / unified files endpoint (#1398).
- (B) — add `cover_file_id UUID NULL REFERENCES files(id) ON DELETE SET NULL` on `commodities`. Migration + swag annotation.
- Endpoint extension: `PATCH /commodities/{id}/cover { file_id }` (or fold into `PUT /commodities/{id}` payload).

### Frontend
- `CommodityCard` (in CommoditiesListPage) — render `` if cover/photo available, else current emoji. Aspect-ratio'd to the existing card slot.
- Sheet preview (`CommodityPreview`) — same render in the header slot.
- Detail page — full-width hero image; Files tab gains a "star" affordance per photo to set cover.
- Loading state: skeleton box with the type emoji while the signed URL resolves.
- Failure state: keep current emoji, no broken-image icon.

### Performance / a11y

- Use the already-existing thumbnail signed-URL pipeline (`go/services/thumbnail_generation_worker.go`) — never serve full-size on cards.
- ``, `loading="lazy"`, fixed dimensions to avoid CLS.
- Bundle-size: no new dep; render uses native ``.

## Acceptance criteria

- [ ] Research sketch posted in this issue + chosen approach noted.
- [ ] If (B) ships: migration + swag + endpoint + frontend toggle + tests.
- [ ] List card / Sheet preview / detail header all show the cover photo when present, with a sensible fallback chain (cover → first photo → type emoji).
- [ ] Component tests cover all three states; jest-axe clean.
- [ ] E2E: upload photo → see it on the card; clear cover → fallback works.

## Cross-refs

- Files unification: #1398 / #1399 (decides where photos live in the unified files surface).
- List page: #1410 (PR #1447) — current emoji-only render.
- Parent epic: #1397.
- Source: `inventario-design` mock — ItemCard, ItemDetail hero image.

### PR DESCRIPTION
**Pull request by denisvmedia**
Closes #1451.

Implements all three options from the issue body in one PR — option (A) auto-pick (first photo by `created_at`) as the default, option (B) `cover_file_id` as the explicit override, option (C) "Use as cover photo?" checkbox in the upload dialog as polish. Resolution order at read time is `cover_file_id` → first photo → type emoji.

## Backend

- **Migration**: `commodities.cover_file_id` (TEXT, FK `files(id)` ON DELETE SET NULL). Generated via `bash scripts/generate-migration.sh add_commodity_cover_file_id`. Manual `ON DELETE SET NULL` fix-up because the Ptah generator still doesn't emit delete behaviour from `on_delete="SET NULL"` annotations — same pattern as `fk_user_default_group` (PR #1306).
- **Resolver**: new `services.CommodityCoverService` resolves the cover for one or many commodities. Stale or cross-commodity overrides fall back to the auto-pick path so the slot never blanks out on a deleted-image race.
- **Response shape**: covers surface under `meta.covers[id]` on the list response and `meta.cover` on single-commodity GET / PATCH responses. Each entry carries `file_id`, signed thumbnail URLs (`small`/`medium`) and `source` (`first_photo` | `explicit`).
- **New endpoint**: `PATCH /commodities/{id}/cover` accepts `{ data: { type: "commodity_cover", attributes: { file_id } } }`; null/empty clears the override. Validates the file already belongs to this commodity and is an image (422 otherwise).
- **Tests**: 8 cover-service unit tests (auto-pick by `created_at` ASC, batch, non-image defensive guard, empty-user short-circuit, explicit-preferred, stale-fallback, cross-commodity rejected) plus 4 apiserver integration tests (list-with-cover-meta, set+clear round-trip, foreign-file rejection, non-image rejection). The apiserver test fixture now sets `FileSigningKey` so signed URLs work under the in-memory registry.

## Frontend

- **`CommodityThumb`**: new component, renders cover with fallback chain `cover → first photo → type emoji`. Auto-picks `small` (≤150px) vs `medium` (300px) thumbnail variant; lazy-loaded `` with fixed dims; falls back to the emoji on `onError`. Replaces the three emoji boxes on the items list (card / row / Sheet preview) and the size-12 emoji block on the detail-page hero.
- **Star toggle** on each photo file in `EntityFilesPanel` / `FileCard`: filled star = explicit cover, outline star = first-photo auto-pick, hover-revealed outline = "set as cover". Click the filled star to clear the override; click any other star to set explicit. `useSetCommodityCover(id)` patches the detail cache and invalidates the list cache so card / preview / hero update instantly.
- **Upload-time checkbox**: "Use as cover photo?" appears in the upload dialog metadata step, shown only when the upload is commodity-linked AND the parent supplied `onSetCover`. Defaults ON for the first photo when the commodity has no current cover; OFF otherwise. Switching category away from photos clears the flag. After upload, the first uploaded item with the flag is promoted via the cover endpoint.
- **Tests**: 8 `CommodityThumb` component tests (cover, fallback, error, variant pick, alt safety net, axe-clean), 1 list-page integration test (cover meta → image vs absent → emoji), 8 new `FileCard` star-toggle tests, 5 new upload-dialog cover-checkbox tests (hidden without handler, default ON for first photo, OFF when cover exists, cleared on category change, promotion after upload).

## OpenAPI / codegen

Regenerated `docs/swagger.{json,yaml}` + `docs.go` and `frontend/src/types/api.d.ts` via `make swagger`. Swagger route-coverage gate from #1442 passes.

## Test plan

- [x] `go test ./...` — green (cover service + apiserver + jsonapi + 8 other packages)
- [x] `golangci-lint run ./...` — 0 issues
- [x] `npm run typecheck` — clean
- [x] `npm run lint` — clean
- [x] `npx vitest run` — 514 / 514 pass (was 502; +12 new tests)
- [ ] Manual smoke: upload a photo on the detail page → see thumbnail on card, in Sheet preview, on hero. Click star on a different photo → cover flips. Click filled star → fallback to first photo. Delete the cover photo → fallback to next earliest photo. Drop a third photo without ticking "Use as cover" → cover stays where it was.
- [ ] E2E (deferred): adding the `upload photo → see on card → clear → fallback` spec would touch `e2e/tests/file-uploads.spec.ts` / `includes/uploads.ts`, which is the active surface for #1474. Ship that spec on top of #1474 once the helper port lands.

## Notes for review

- The N+1 cover lookup (one `ListByLinkedEntityAndMeta` per commodity in the list path) is fine for a 50-per-page card view but worth a single batched `SELECT DISTINCT ON (linked_entity_id) ...` once we hit a perf budget. Adding a registry method requires touching memory + postgres + interface; deliberately deferred.
- `EntityFilesPanel` / `FileCard` got cover-toggle props but they remain optional, so non-commodity callers (the global Files page, location detail, future entities) keep their current read-only behaviour.

### CODE REVIEW

File: go/schema/migrations/_sqldata/1777918262_add_commodity_cover_file_id.up.sql [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+-- ALTER statements: --
+ALTER TABLE commodities ADD COLUMN cover_file_id TEXT;
+-- Add foreign key constraints for table: commodities --
+-- ALTER statements: --
+-- ON DELETE SET NULL is added manually: the Ptah generator does not yet emit
+-- delete behaviour even when the model carries on_delete="SET NULL". Keeping
+-- it explicit here ensures that deleting a file silently clears the cover
+-- override (the resolver's first-photo path takes over) instead of blocking
+-- the file delete.
+ALTER TABLE commodities ADD CONSTRAINT fk_commodity_cover_file FOREIGN KEY (cover_file_id) REFERENCES files(id) ON DELETE SET NULL;
```
[the remark this answers is not in the document]
Comment (denisvmedia) [reply go/schema/migrations/_sqldata/1777918262_add_commodity_cover_file_id.up.sql:15]: The migration *was* generated via `bash scripts/generate-migration.sh add_commodity_cover_file_id` — the Ptah generator just doesn't emit `ON DELETE` behaviour from `on_delete="SET NULL"` annotations yet (a known gap, called out in `1776710604_add_user_default_group_id.up.sql` with the same manual fix-up). Without the `SET NULL` clause, deleting a photo would block on the FK; this PR depends on the resolver's "stale override falls back" path which only works because the column actually nulls itself when the file row goes away.

I'd rather not regress that behaviour to silence the lint, but I agree the right fix is upstream. Filed/​tracked as a follow-up against the migrator. Until then, the convention from #1306 — generate then add the `ON DELETE` line manually with a comment — keeps the migration honest and the diff transparent. Resolving as "won't change in this PR".

File: go/apiserver/commodities.go [hunk 2]
```diff
@@ -317,6 +379,133 @@ func (api *commoditiesAPI) updateCommodity(w http.ResponseWriter, r *http.Reques
...
+		desired = strings.TrimSpace(*input.Data.Attributes.FileID)
+	}
+
+	if desired != "" {
+		if err := api.validateCoverFile(r.Context(), registrySet.FileRegistry, commodity.ID, desired); err != nil {
+			// Validation errors (`validation.Errors`) are user-facing
+			// 422s; the registry NotFound from a missing file id is too.
+			// renderEntityError treats `validation.Errors` as a generic
+			// internal error, so we surface 422 directly here.
+			unprocessableEntityError(w, r, err)
```
[the remark this answers is not in the document]
Comment (denisvmedia) [reply go/apiserver/commodities.go]: Fixed in f018ed3. The handler now splits its error mapping:

- `validation.Errors` → `unprocessableEntityError` (422), as before — it's a user-input issue and `renderEntityError`'s default branch would log it as a 500.
- Anything else → `renderEntityError`, which maps `registry.ErrNotFound` → 404 per the swagger contract and unknown registry failures → 500.

`validateCoverFile` also returns the underlying `registry.Get` error verbatim now (instead of squashing every error to `ErrNotFound`) so transient registry failures aren't masked. New test `TestCommodityCoverPatch_NotFoundForUnknownFile` asserts the 404 path.

File: go/apiserver/commodities.go [hunk 3]
```diff
@@ -317,6 +379,133 @@ func (api *commoditiesAPI) updateCommodity(w http.ResponseWriter, r *http.Reques
...
+	}
+	if file == nil || file.File == nil {
+		return registry.ErrNotFound
+	}
+	if file.LinkedEntityType != "commodity" || file.LinkedEntityID != commodityID {
+		return validationError("file_id", "file is not attached to this commodity")
+	}
+	if file.Type != models.FileTypeImage {
+		return validationError("file_id", "file is not an image")
+	}
```
[the remark this answers is not in the document]
Comment (denisvmedia) [reply go/apiserver/commodities.go:517]: Fixed in f018ed3. `validateCoverFile` now also enforces `file.Category == FileCategoryPhotos`. `LinkedEntityMeta=="images"` is implicit because the file must already be linked to this commodity and that meta value is the only one paired with `Category=photos` for commodity attachments (per the validator on `models.FileEntity`).

Test `TestCommodityCoverPatch_RejectsImageInWrongCategory` covers the JPEG-as-invoice case explicitly.

File: go/services/commodity_cover_service.go [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+	}
+	if first == nil || first.File == nil {
+		return nil, false
+	}
+	// Skip non-image rows — the legacy "images" bucket is officially
+	// images-only but a defensive guard avoids surfacing the wrong file
+	// on a card if a caller mis-classified an upload.
+	if first.Type != models.FileTypeImage {
+		return nil, false
+	}
```
[the remark this answers is not in the document]
Comment (denisvmedia) [reply go/services/commodity_cover_service.go]: Good catch — fixed in f018ed3. `pickFirstPhoto` now filters non-eligible rows *before* the earliest-by-`CreatedAt` pick, so a mis-classified row at the top of the bucket can no longer shadow a valid photo uploaded a second later. The eligibility check was extracted to `isCoverEligible` (used by both the read and the write path) — both `Type=image` and `Category=photos` must hold.

File: go/services/commodity_cover_service.go [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+	if err != nil || file == nil || file.File == nil {
+		return nil
+	}
+	if file.LinkedEntityType != "commodity" || file.LinkedEntityID != commodityID {
+		return nil
+	}
+	if file.Type != models.FileTypeImage {
+		return nil
+	}
+	return file
```
[the remark this answers is not in the document]
Comment (denisvmedia) [reply go/services/commodity_cover_service.go:173]: Fixed in f018ed3. `fetchExplicitCover` now goes through the same `isCoverEligible` predicate as the auto-pick path, so explicit overrides have to satisfy `Type=image` AND `Category=photos`. A stale override that fails this check falls through to the first-photo path (existing `TestCommodityCoverService_StaleOverrideFallsBack` covers that branch).

File: go/services/commodity_cover_service.go [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+	Thumbnails  map[string]string
+	Source      CoverSource
+}
+
+// CommodityCoverService resolves the cover image for one or more
+// commodities. The current strategy is option (A): the earliest file
+// (by `created_at`) with `linked_entity_type=commodity` /
+// `linked_entity_id=<id>` / `category=photos`. Issue #1451 option (B)
+// — `cover_file_id` override — slots in here once the migration lands.
+type CommodityCoverService struct {
```
[the remark this answers is not in the document]
Comment (denisvmedia) [reply go/services/commodity_cover_service.go]: Fixed in f018ed3. The doc comments on `CoverSource` and `CommodityCoverService` now describe the actual two-step resolution order (explicit → first-photo) shipping in this PR instead of pointing at a future migration.

File: frontend/src/pages/commodities/CommodityDetailPage.tsx [hunk 7]
```diff
@@ -347,6 +385,20 @@ export function CommodityDetailPage() {
           name: commodity?.name,
         }}
         initialFiles={pendingDropFiles}
+        // Issue #1451 option C — promote the first uploaded photo to
+        // the cover when the commodity has no explicit cover yet. The
+        // checkbox in the metadata step defaults to ON for that case.
+        commodityHasCover={!!commodity?.cover_file_id}
+        onSetCover={(fileId) => {
```
[the remark this answers is not in the document]
Comment (denisvmedia) [reply frontend/src/pages/commodities/CommodityDetailPage.tsx:399]: Fixed in f018ed3. `commodityHasCover` now reads `!!commodity?.cover` — i.e. "the BE returned a resolved cover under `meta.cover`", which covers both the explicit override and the auto-pick first-photo. So uploading a new photo on a commodity that already shows the type emoji as a cover (no `cover_file_id`, no photos at all) still defaults the checkbox ON, but uploading on a commodity that has any auto-pick cover leaves it OFF as expected.

File: frontend/src/components/files/UploadFilesDialog.tsx [hunk 8]
```diff
@@ -405,50 +499,91 @@ function MetadataStep({ items, onPatch }: MetadataStepProps) {
...
+  return items.map((it) => {
+    if (promoted) return it
+    if (it.category !== "photos") return it
+    promoted = true
+    return { ...it, useAsCover: true }
+  })
+}
+
 interface ProgressStepProps {
   items: FileItem[]
```
[the remark this answers is not in the document]
Comment (denisvmedia) [reply frontend/src/components/files/UploadFilesDialog.tsx]: Fixed in f018ed3. New `coverTouched` state flips to `true` on the first `setCoverFlag` call and resets in `reset()` on dialog close. `applyCoverDefaults` short-circuits when it's true, so unchecking the default and then adding/​removing/​re-categorising another file no longer silently re-promotes the first photo. New test (`does not re-promote a cover after the user explicitly unchecks it and adds another file`) walks through the exact scenario from the comment.
